[release/tune] test_bookkeeping_overhead fails (too much overhead)

The release tests fails with the following output:

```
2021-03-29 07:17:55,040 INFO tune.py:549 -- Total run time: 3520.55 seconds (3517.42 seconds for the tuning loop).
Traceback (most recent call last):
  File "workloads/test_bookkeeping_overhead.py", line 42, in <module>
    main()
  File "workloads/test_bookkeeping_overhead.py", line 38, in main
    max_runtime=max_runtime)
  File "/home/ray/release-automation-tune_scalability_tests/workloads/_trainable.py", line 145, in timed_tune_run
    f"The {name} test took {time_taken:.2f} seconds, but should not " \
AssertionError: The bookkeeping overhead test took 3553.69 seconds, but should not have exceeded 800.00 seconds. Test failed. 

--- FAILED: BOOKKEEPING OVERHEAD ::: 3553.69 > 800.00 ---
```

This is a regression from the last release, where this test finished in under 750 seconds.

--- a/doc/source/tune/user-guide.rst
+++ b/doc/source/tune/user-guide.rst
@@ -818,6 +818,11 @@ These are the environment variables Ray Tune currently considers:
 * **TUNE_PLACEMENT_GROUP_PREFIX**: Prefix for placement groups created by Ray Tune. This prefix is used
   e.g. to identify placement groups that should be cleaned up on start/stop of the tuning run. This is
   initialized to a unique name at the start of the first run.
+* **TUNE_PLACEMENT_GROUP_RECON_INTERVAL**: How often to reconcile placement groups. Reconcilation is
+  used to make sure that the number of requested placement groups and pending/running trials are in sync.
+  In normal circumstances these shouldn't differ anyway, but reconcilation makes sure to capture cases when
+  placement groups are manually destroyed. Reconcilation doesn't take much time, but it can add up when
+  running a large number of short trials. Defaults to every ``5`` (seconds).
 * **TUNE_PLACEMENT_GROUP_WAIT_S**: Default time the trial executor waits for placement
   groups to be placed before continuing the tuning loop. Setting this to a float
   will block for that many seconds. This is mostly used for testing purposes. Defaults

--- a/python/ray/tune/ray_trial_executor.py
+++ b/python/ray/tune/ray_trial_executor.py
@@ -1005,8 +1005,9 @@ class RayTrialExecutor(TrialExecutor):
             self._update_avail_resources()
             return self._avail_resources.gpu > 0
 
-    def cleanup(self):
+    def cleanup(self, trial_runner):
         self._trial_cleanup.cleanup(partial=False)
+        self._pg_manager.reconcile_placement_groups(trial_runner.get_trials())
         self._pg_manager.cleanup(force=True)
         self._pg_manager.cleanup_existing_pg(block=True)
 

--- a/python/ray/tune/ray_trial_executor.py
+++ b/python/ray/tune/ray_trial_executor.py
@@ -180,6 +180,10 @@ class RayTrialExecutor(TrialExecutor):
         if self._wait_for_pg < 0:
             self._wait_for_pg = None
 
+        self.last_pg_recon = 0
+        self.pg_recon_interval = float(
+            os.environ.get("TUNE_PLACEMENT_GROUP_RECON_INTERVAL", "5"))
+
         self._buffer_length = int(os.getenv("TUNE_RESULT_BUFFER_LENGTH", 1000))
         self._buffer_min_time_s = float(
             os.getenv("TUNE_RESULT_BUFFER_MIN_TIME_S", 0.))
@@ -893,7 +897,13 @@ class RayTrialExecutor(TrialExecutor):
     def on_step_end(self, trial_runner):
         self._just_staged_trials.clear()
 
-        self._pg_manager.reconcile_placement_groups(trial_runner.get_trials())
+        if time.time() > self.last_pg_recon + self.pg_recon_interval:
+            # Only do this every now and then - usually the placement groups
+            # should not get out of sync, and calling this often is inefficient
+            self._pg_manager.reconcile_placement_groups(
+                trial_runner.get_trials())
+            self.last_pg_recon = time.time()
+
         self._pg_manager.cleanup()
 
     def save(self, trial, storage=Checkpoint.PERSISTENT, result=None):

--- a/python/ray/tune/tests/test_trial_runner_pg.py
+++ b/python/ray/tune/tests/test_trial_runner_pg.py
@@ -69,6 +69,9 @@ class TrialRunnerPlacementGroupTest(unittest.TestCase):
 
         Eventually they should be scheduled sequentially (i.e. in pairs
         of two)."""
+        # Since we check per-step placement groups, set the reconcilation
+        # interval to 0
+        os.environ["TUNE_PLACEMENT_GROUP_RECON_INTERVAL"] = "0"
 
         def train(config):
             time.sleep(1)

--- a/python/ray/tune/trial_executor.py
+++ b/python/ray/tune/trial_executor.py
@@ -278,7 +278,7 @@ class TrialExecutor:
         """Returns True if GPUs are detected on the cluster."""
         return None
 
-    def cleanup(self, trial):
+    def cleanup(self, trial_runner):
         """Ensures that trials are cleaned up after stopping."""
         pass
 

--- a/python/ray/tune/trial_runner.py
+++ b/python/ray/tune/trial_runner.py
@@ -1144,7 +1144,7 @@ class TrialRunner:
         self.trial_executor.stop_trial(trial, error=error, error_msg=error_msg)
 
     def cleanup_trials(self):
-        self.trial_executor.cleanup()
+        self.trial_executor.cleanup(self)
 
     def __getstate__(self):
         """Gets state for trial.

--- a/rllib/tests/test_placement_groups.py
+++ b/rllib/tests/test_placement_groups.py
@@ -1,3 +1,4 @@
+import os
 import unittest
 
 import ray
@@ -52,6 +53,7 @@ class _TestCallback(Callback):
 
 class TestPlacementGroups(unittest.TestCase):
     def setUp(self) -> None:
+        os.environ["TUNE_PLACEMENT_GROUP_RECON_INTERVAL"] = "1"
         ray.init(num_cpus=6)
 
     def tearDown(self) -> None:

--- a/rllib/tests/test_placement_groups.py
+++ b/rllib/tests/test_placement_groups.py
@@ -53,7 +53,7 @@ class _TestCallback(Callback):
 
 class TestPlacementGroups(unittest.TestCase):
     def setUp(self) -> None:
-        os.environ["TUNE_PLACEMENT_GROUP_RECON_INTERVAL"] = "1"
+        os.environ["TUNE_PLACEMENT_GROUP_RECON_INTERVAL"] = "0"
         ray.init(num_cpus=6)
 
     def tearDown(self) -> None:

